feat: HomewView 구현

parkyounbae opened the pull request:
프로필, 이름, 커밋카운트, 커밋 Streak, 잔디, 새로고침 버튼, 설정 버튼 구현
정보를 받아오고 만약 정보가 없다면 빈 잔디와 카운트 숫자에는 0 표시, username라벨 탭시 설정 화면으로 넘어감
화면이 바뀔때 마다 정보가 업데이트 되도록 설정

Resolves: GITIT-47

parkyounbae committed:
feat: HomewView 구현

프로필, 이름, 커밋카운트, 커밋 Streak, 잔디, 새로고침 버튼, 설정 버튼 구현
정보를 받아오고 만약 정보가 없다면 빈 잔디와 카운트 숫자에는 0 표시, username라벨 탭시 설정 화면으로 넘어감
화면이 바뀔때 마다 정보가 업데이트 되도록 설정

Resolves: GITIT-47

parkyounbae committed:
refactor: OperationQueue대신에 DispatchQueue 를 사용

간단한 작업에는 더 유리한 DispatchQueue 사용

Resolve: GITIT-47

parkyounbae committed:
test: test code 주석처리

테스트 코드 주석처리

Resolve: GITIT-47

parkyounbae committed:
refact: settingView에서 뒤로가기 버튼을 눌려야 업데이트 하게 수정

원래 뷰가 로드 될때마다 데이터가 업데이트 되었지만 이를 백버튼 눌렀을때만 작동하도록 수정
정보 업데이트하는 함수를 따로 만듦

Resolve: GITIT-47

parkyounbae committed:
test: 테스트 코드 주석처리

테스트 코드 주석처리

Resolve: GITIT-47

parkyounbae committed:
feat: commit streak 에 days 문구 추가

days 문구 추가

Resolve: GITIT-47

sweethoneybee committed:
main 변경사항 GITIT-47-HomeView에 반영

parkyounbae committed:
refactor: 이미지 부분 캐싱 이용

홈 화면 이미지 캐싱 이용

Resolve: GITIT-47

sweethoneybee commented on Git-it/ViewController/HomeViewController.swift:
```diff
@@ -0,0 +1,354 @@
+    }
+    
+    func addGrassCollectionView() {
+        grassCollectionView = GrassCollectionView(frame: CGRect.zero, collectionViewLayout: UICollectionViewFlowLayout.init())
```
여기랑 collectionViewCellFlowLayout 함수를 합쳐도 될 듯
GrassCollectionView 생성할 때 인자로 collectionViewCellFlowLayout 함수의 리턴값을 넘겨주면
코드를 더 깔끔히 유지할 수 있을 듯
그리고 레이아웃 설정하는 함수에서 스크롤해주는 부분을 여기 함수로 빼오면 의미가 더 적절할 듯.

sweethoneybee commented on Git-it/ViewController/HomeViewController.swift:
```diff
@@ -0,0 +1,354 @@
+        flowLayout = UICollectionViewFlowLayout()
+        flowLayout.minimumInteritemSpacing = 2
+        flowLayout.minimumLineSpacing = 2
+        flowLayout.scrollDirection = .horizontal
```
여기서 flowLayout을 리턴해주는 방식으로 수정하는 걸 추천.  
그러면 `GrassCollectionView(frame: CGRect.zero, collectionViewLayout: collectionViewFlowLayout())`식으로 
콜렉션 뷰를 생성할 수 있을 것임.  
그러면 scrollToItem 메소드도 addGrassCollectionView 함수로 옮기는 게 적절함

sweethoneybee commented on Git-it/ViewController/HomeViewController.swift:
```diff
@@ -0,0 +1,354 @@
+        
+//        self.userData = try? JSONDecoder().decode(CommitsSummary.self, from: GitItApi.commitsSummary(UserInfo.username!).sampleData)
+        
+        DispatchQueue.main.async {
```
여기는 UI작업을 하지 않으니 메인큐 안쓰셔도 되고요
지금 네트워크클래스가 기본 URLSession 클래스를 사용하고 있어서 데이터 페칭도
자동으로 백그라운드큐에서 진행되니, 백그라운드큐에도 안 넣어도 잘 작동할 거예요

sweethoneybee commented on Git-it/ViewController/HomeViewController.swift:
```diff
@@ -0,0 +1,354 @@
+                case .failure(let error):
+                    print(error)
+                case .success(let commitSummary):
+                    self.userData = commitSummary
```
다만 이제 여기서 `updateView()` 함수를 호출하게 된다면
이거는 메인큐로 감싸 줘야합니다. UI관련 작업이니깐.

sweethoneybee commented on Git-it/ViewController/HomeViewController.swift:
```diff
@@ -0,0 +1,354 @@
+        }
+    }
+    
+    func updateView() {
```
updateView는 updateUserData가 성공적으로 끝났을 때 호출되는 게 적절해보여요
현재 이 함수는 updateUserData가 호출되고 데이터를 페칭해오는 동안
이미 UI갱신을 끝내버립니다(updateUserData는 비동기적으로 데이터를 갱신하기 때문)
그래서 갱신된 데이터를 반영하지 못하고 있어요

updateView 함수는 updateUserData가 데이터를 성공적으로 받아온 경우 호출되면 좋을 듯 합니다
또 받아온 이미지 url(userData의 url)과 UserInfo의 url을 비교해서 다르면 UserInfo를 갱신하고
ImageCache를 통해 새로운 url의 이미지를 로드해서 UI를 갱신해주면 좋을 듯해요

sweethoneybee commented on Git-it/ViewController/HomeViewController.swift:
```diff
@@ -0,0 +1,354 @@
+    }
+    
+    @IBAction func touchUpRefreshButton(_ sender: UIButton) {
+        updateView()
```
위의 updateView 와 마찬가지의 이유로 여기서는 데이터 페칭을 하는 게 더 적절해보입니다

sweethoneybee reviewed (commented):
구현 고생하셨어요. 특히 UI까지 같이 하신다고 고생하셨습니다 ㅠㅠ.
뭔가 옵셔널로 범벅이 되다보니 UI하는데 귀찮음이 더 많았을 것 같은데, 이거는 뭔가 더 좋은 방법이 있을 것 같아요(지금 당장 떠오르는 건 없지만). 옵셔널 범벅이 정말 옳은 건지 다시 고민하게 됐습니다 ㅋㅋㅋ.

비동기처리와 관련해서 로직상 수정해야할 부분이랑 콜렉션뷰 레이아웃 함수와 관련해서 정도만 피드백 된 것 같아요
콜렉션뷰는 작동상 문제는 없지만, 비동기처리는 작동상 문제가 있으니 꼭 잘 수정해보시기 바랍니다.

parkyounbae committed:
refactor: 뷰 업데이트 방식 변경

뷰 업데이트 함수를 데이터 패치 이후에 작동하도록 변경
콜렉션 뷰 플로우레이아웃을 반환값을 통해 할당
뷰 업데이트시 userinfo에 있던 프로필 이미지 url도 같이 업데이트
새로고침시 유저 정보를 업데이트

Resolve: GITIT-47
